File: challenge-02-dart/server/lib/store.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import './shortened_url.dart';

class Store {
  static final Store _store = Store._internal();

  factory Store() {
    return _store;
  }

  Store._internal();

  String _name = 'Store';

  String get name => _name;

  // Key value list of shortened urls
  Map<String, ShortenedURL> _urls = {};

  ShortenedURL? getURL(String key) {
    return _urls[key];
  }

  ShortenedURL addURL(String url) {
    String key = _generateHashKey(url);
    if (_urls.containsKey(key)) {
      return _urls[key]!;
    }
    ShortenedURL shortenedURL =
        ShortenedURL(key, url, 'http://localhost:8080/$key');
    _urls[key] = shortenedURL;
    return shortenedURL;
  }

  String _generateHashKey(String url) {
    var bytes = utf8.encode(url);
    var digest = sha1.convert(bytes);
    return digest.toString().substring(0, 5);
  }
}

File: challenge-02/server/store.ts
import { ShortenedURL } from '@/shared/types';
import crypto from 'crypto';

const generateHashKey = (url: string) => {
  const hash = crypto.createHash('sha1');
  hash.update(url);
  return hash.digest('hex').slice(0, 5);
}
const urls = new Map<string, ShortenedURL>();

export const addUrl = (url: string): ShortenedURL => {
  const key = generateHashKey(url);
  if (urls.has(key)) {
    return urls.get(key) as ShortenedURL;
  }

  const shortenedUrl: ShortenedURL = {
    key,
    url,
    shortURL: `http://localhost:3000/${key}`,
  };

  urls.set(key, shortenedUrl);

  return shortenedUrl;
}

export const getURL = (key: string): ShortenedURL | undefined => {
  return urls.get(key);

}
